lib/SIL/Utils/OwnershipLiveness.cpp

```diff
@@ -38,6 +38,13 @@ struct LinearLivenessVisitor :
     linearLiveness(linearLiveness){}
 
   bool handleUsePoint(Operand *use, UseLifetimeConstraint useConstraint) {
+    if (isa<ExtendLifetimeInst>(use->getUser())) {
```

Comment:
I'm confused about the `handleUsePoint` contract.

If `extend_liveness` should be passed to `handleUsePoint` as `LifetimeEnding`, then that should be specified in the function interface comment (similar to the comment in the commit log) since that is an inversion of the expected value. Then we should not need additional checks within the visitor implementations of `handleUsePoint.

Preferably, `extend_liveness` would be passed to `handleUsePoint` as `NonLifetimeEnding` to avoid confusion, and the visitor implementations of `handleUsePoint` will check for `extend_liveness` if they need to invert the value. In that case, the explanatory comment in the commit message needs to be on each visitor implementation that inverts the sense of `extend_lifetime`.

Instead, this diff shows extra checks for `extend_lifetime` on both sides of `handleUsePoint` and there aren't any comments to explain away the confusion.

Reply:
That makes sense.  Fixed to pass `NonLifetimeEnding` to `handleUsePoint`.  And actually, is there any reason that we want the `extend_lifetime` instructions to be added to `PrunedLiveness` with `/*lifetimeEnding=*/true`?  I haven't thought of one--we still get all users within the `LinearLifetime` boundary either way--so I removed that special behavior.

Reply:
> is there any reason that we want the extend_lifetime instructions to be added to PrunedLiveness with /*lifetimeEnding=*/true? I haven't thought of one--we still get all users within the LinearLifetime boundary either way--so I removed that special behavior.

Oh, wonderful. I just assumed that you needed to invert the `lifetimeEnding` value in PrunedLiveness to avoid special cases elsewhere. It's always better if the code doesn't need to lie about anything though.